datafusion/physical-expr/src/expressions/binary.rs

```diff
@@ -649,6 +650,30 @@ impl PhysicalExpr for BinaryExpr {
             right: Arc::clone(self.right()),
         })
     }
+
+    fn children(&self) -> Vec<Arc<dyn PhysicalExpr>> {
+        vec![self.left.clone(), self.right.clone()]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        children: Vec<Arc<dyn PhysicalExpr>>,
+    ) -> Result<Arc<dyn PhysicalExpr>> {
+        Ok(Arc::new(BinaryExpr::new(
```

Comment:
I suggest we should also `assert_eq!(children.len(), 2)` to catch bugs earlier

Reply:
I already add the length check in the method `with_new_children_if_necessary()`, and expr.with_new_children() should not be called directly. This is also aligned with the ExecutionPlan.

````
pub fn with_new_children_if_necessary(
    expr: Arc<dyn PhysicalExpr>,
    children: Vec<Arc<dyn PhysicalExpr>>,
) -> Result<Arc<dyn PhysicalExpr>> {
    if children.len() != expr.children().len() {
        Err(DataFusionError::Internal(
            "PhysicalExpr: Wrong number of children".to_string(),
        ))
    } else if children.is_empty()
        || children
            .iter()
            .zip(expr.children().iter())
            .any(|(c1, c2)| !Arc::ptr_eq(c1, c2))
    {
        expr.with_new_children(children)
    } else {
        Ok(expr)
    }
}

````